### ISSUE TITLE
Mock DB

### ISSUE THREAD

**Issue by ronniekritou** [rethinkdb/rethinkdb-go#264]
Not so much of an Issue but more of a question. Trying to find the best approach for my unit tests, currently have a rethinkDB instance running on a DO node and I just drop tables after each test and recreate them. Wondering if there is a better approach then what I am currently doing? 

Thanks!

**Comment by thinkong** [comment]
r.db('mockdb').table('mocktable').delete() will just delete all the data.. if that was what you needed.
Not exactly sure what you are trying to do though

**Comment by ronniekritou** [comment]
Thanks for the reply, thats what i'm currently doing now. But I was looking if there was a way to mock similar to https://github.com/DATA-DOG/go-sqlmock

**Comment by kofalt** [comment]
It sounds like you're asking for an entire reimplementation of the rethinkdb logic & behaviour that stores state on memory (probably in-process) instead of disk. If I'm understanding you correctly, this is firmly outside the scope of gorethink, but there's a couple things you can do instead:

1) Nothing - Rethink [intelligently holds state in memory](https://www.rethinkdb.com/docs/memory-usage/) already, so I would assert that the primary difference any optimization could make would mainly be to write latency. Depending on your test workload, there may not be a compelling difference in your observed speed.

2) Run Rethink from a ramdisk - `mount -t tmpfs rethink /some/folder`, run rethink with a data directory inside that. This should largely eliminate write latency.

3) Implement - If you did want to go full steam ahead on a mock driver, I suspect implementing a [Session](https://godoc.org/github.com/dancannon/gorethink#Session) object would probably be a good place to start, with methods like `(s *Session) Exec(q Query) error`. I bet folks would be interested if you chose to go that path.

Hopes this helps!

**Comment by ronniekritou** [comment]
Might end up going with option # 3 to help out the community. Thanks for the reply though!

**Comment by dancannon** [comment]
Sorry for taking so long to reply, I wanted to play around with a possible solution before replying. 

As @kofalt mentioned one solution would be to reimplement all of RethinkDBs logic however [reqlite](github.com/neumino/reqlite) has already done that so I attempted to create a go binding for this package with go-duktape however they didnt seem to compatible. I think the 3rd option that @kofalt mentioned is probable a better solution for now.

@ronniekritou if you could have a go at implementing something I would really appreciate that! If you need any help or have any questions just let me know, I am usually online on the RethinkDB slack.

**Comment by ereyes01** [comment]
I'm currently hiding all ReQL / RethinkDB operations behind my own interface, then mocking the interface in tests. I'll first take a feeble stab at how one might design an integrated gorethink mock, and then I'll describe my current approach in more detail.

Creating a mock of gorethink's ReQL implementation looks kind of tricky on the surface to uninformed me. I think a (likely naive) framework for how I'd approach / think about it would be:
- Devise a way to spec an expected ReQL expression. It looks like ReqL expressions reduce to a protobuf under the covers. Maybe initialize the mock with some expected ReQL expression and save the protobuf as the expected outcome, as well as the expected form of execution (Run / RunWrite / Exec).
- Also as part of the mock's initialization, save a value to be sent over the mock cursor, or a mock response if it's a write operation.
- Open up some way to inject the mock into a session object. This could possibly be left up to the caller to inject a mock session object into their code themselves.
- The mock should have some `Verify()` method that the test should call to ensure the expected query was given, and that the response / cursor was consumed.

...

In the spirit of "getting shit done" I simply devised my own way to test my code that depends on gorethink. As I explained above, I designed my own interface that defines a method for each database operation my code will ever do at quite a fine granularity. My code only ever interacts with the DB via this interface.

At test time, I inject a mock instance of my interface (typically 1 mock implementation for each test) that's booby trapped to blow up the test unless it's called in the exact way the test expects. Here's an example of what such an interface could look like:

``` go
// DBConnection represents a connection to the DB used by the API server
type DBConnection interface {
    // ConnectToDB initializes the connection to the database and saves the
    // needed state.
    ConnectToDB(config *Config) error

    // GetByID retrieves an object by its unique ID number (i.e. its primary
    // key)
    GetByID(table, ID string) (DBObject, error)

    // DoQuery looks up one of more objects according to the given query..
    DoQuery(query *Query) ([]DBObject, error)

    // InsertObject adds a new object into the named table of the DB.
    InsertObject(table string, object DBObject) error
}
```

A mock for a test that expects to retrieve a document by primary key might look like:

(Uses github.com/onsi/gomega for assertions):

``` go
type mockByID struct {
    gotID        bool
    expectedTable string
    expectedID   string
    object       DBObject
}

func (m *mockByID) ConnectToDB(config *Config) error {
    Expect(false).To(BeTrue(), "shouldn't be called in this test")
    return nil
}

func (m *mockByID) InsertObject(table string, object DBObject) error {
    Expect(false).To(BeTrue(), "shouldn't insert object in this test")
    return nil
}

func (m *mockByID) DoQuery(query *Query, result interface{}) error {
    Expect(false).To(BeTrue(), "shouldn't query in this test")
    return nil
}

func (m *mockByID) GetByID(table, ID string) (DBObject, error) {
    defer func() { m.gotID = true }()
    Expect(m.gotID).To(BeFalse(), "Expect only 1 get-by-id per request")

    Expect(table).To(Equal(m.expectedTable))
    Expect(ID).To(Equal(m.expectedID))

    return m.object, nil
}

func (m *mockByID) verifyCalled() {
    Expect(m.gotID).To(BeTrue(), "Expected to get object by ID")
}
```

Go interfaces are nifty constructs, and it so happens you can use this approach to mock out just about any external dependency you rely on in Go. The downside of such an approach is that it's not as comprehensive as a truly integrated ReQL mock could achieve. While this mock interface injection helps you spec out and test your DB operations at a logical level, it does not help you spec out the concrete ReQL expressions you will need to execute against the DB. Also, it lends itself well to modeling simple ReQL expressions, but is perhaps more unwieldy to spec really complex ReQL expressions. However, if your ReQL is getting that complex, you might want to check yourself and try to simplify your design :laughing: 

Hope this was helpful, thanks for reading.

**Comment by dancannon** [comment]
Thanks for the great write up @ereyes01! In the past I have used a similar solution as you to test code which does not have mocking built-in and I would recommend anybody else do the same until I find the time to work on this issue.

To answer a couple of the points in your comment GoRethink actually sends queries to RethinkDB as JSON which should be relatively easy to mock. I think the major issue will be mocking out the session somehow and providing a good way of replicating the interaction between the `Cursor` and the responses from RethinkDB. I think for the first version this should be as simple as possible and we work from there.

**Comment by dancannon** [comment]
I had a go at implementing mocking in #337, if anybody has the time to have a look at the PR I would really appreciate it!

**Comment by dancannon** [comment]
Released mocking in v2.1.0
